File: lab06/py/IntAPTab.py
from tabulate import tabulate
from math import isqrt
from collections import defaultdict
from typing import List, Dict, Tuple

def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return a

def inverse(a: int, n: int) -> int:
    _, y = euclidean_algorithm(n, a % n)
    return y

def euclidean_algorithm(a: int, b: int) -> Tuple[int, int]:
    big, small = (a, b) if a > b else (b, a)
    prev_x, prev_y = 1, 0
    curr_x, curr_y = 0, 1
    
    while small > 0:
        q = big // small
        big, small = small, big % small
        prev_x, curr_x = curr_x, prev_x - q * curr_x
        prev_y, curr_y = curr_y, prev_y - q * curr_y
    
    return prev_x, prev_y

def chinese_value(a: int, b: int, r1: int, r2: int) -> int | None:
    d = gcd(a, b)
    
    if r1 % d != r2 % d:
        return None
    
    factorization_of_a: Dict[int, int] = factorization(a)
    factorization_of_b: Dict[int, int] = factorization(b)
    
    a_prime, b_prime = a, b
    
    for prime in factorization_of_a.keys():
        if prime in factorization_of_b:
            a_exponent = factorization_of_a[prime]
            b_exponent = factorization_of_b[prime]
            if a_exponent < b_exponent:
                a_prime //= prime ** a_exponent
            else:
                b_prime //= prime ** b_exponent
    
    first_coefficient = inverse(b_prime, a_prime)
    second_coefficient = inverse(a_prime, b_prime)
    
    solution = b_prime * first_coefficient * r1 + a_prime * second_coefficient * r2
    prime = a_prime * b_prime
    
    solution = (solution % prime + prime) % prime
    return solution

def factorization(n: int) -> Dict[int, int]:
    factors = factor(n)

    grouped_factors = defaultdict(int)
    for currentFactor in factors:
        grouped_factors[currentFactor] += 1

    return dict(grouped_factors)

def factor(n: int) -> List[int]:
    result = []

    # Remove powers of 2, 3, 5
    for prime in [2, 3, 5]:
        n = remove_factors(n, result, prime)

    d = 0

    # Trial division with selected forms of d
    square_root = isqrt(n)
    while n > 1 and d <= square_root:
        forms_of_d = [1, 7, 11, 13, 17, 19, 23, 29]

        for current_residue in forms_of_d:
            current_divisor = d + current_residue

            if current_divisor != 1 and n % current_divisor == 0:
                n = remove_factors(n, result, current_divisor)

        d += 30  # Move to the next set of forms

    if n > 1:
        result.append(n)

    return result

def remove_factors(n: int, factors: List[int], prime: int) -> int:
    while n % prime == 0:
        factors.append(prime)
        n //= prime

    return n



def generate_table(a: int, b: int, start_x: int, start_y: int, width: int, height: int):
    modulo_table = []
    
    header = [j for j in range(start_x, start_x + width)]
    
    for i in range(start_y, start_y + height):
        current_row = [i]
        
        for j in range(start_x, start_x + width):
            result = chinese_value(a, b, i, j)
            current_row.append(result)
        
        modulo_table.append(current_row)
    
    return header, modulo_table

def main():
    try:
        a = int(input("Enter number a: "))
        b = int(input("Enter number b: "))
    except ValueError:
        print("Please enter valid positive numbers for a and b.")
        return
    
    start_x, start_y = 0, 0
    width, height = b - start_x, a - start_y
    
    if a <= 0 or b <= 0:
        print("Please enter valid positive numbers for a and b.")
        return
    
    if start_x < 0 or start_x >= b or start_y < 0 or start_y >= a or \
       width <= 0 or height <= 0 or (start_x + width - 1) >= b or (start_y + height - 1) >= a:
        print("Invalid values for startX, startY, width, or height. Please check and try again.")
        return
    
    print(f"Intersection of arithmetic progressions (mod {a}) and (mod {b})")
    headers, table = generate_table(a, b, start_x, start_x, width, height)
    print(tabulate(table, headers=headers, tablefmt="grid"))

main()

File: lab06/cpp/IntApTab.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <iomanip>

using namespace std;

int gcd(int a, int b) {
    while (b) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

pair<int, int> euclidean_algorithm(int a, int b) {
    int big = max(a, b), small = min(a, b);
    int prev_x = 1, prev_y = 0;
    int curr_x = 0, curr_y = 1;

    while (small > 0) {
        int q = big / small;
        int temp = big % small;
        big = small;
        small = temp;
        
        int temp_x = prev_x - q * curr_x;
        prev_x = curr_x;
        curr_x = temp_x;
        
        int temp_y = prev_y - q * curr_y;
        prev_y = curr_y;
        curr_y = temp_y;
    }
    return {prev_x, prev_y};
}

int inverse(int a, int n) {
    auto [_, y] = euclidean_algorithm(n, a % n);
    return y;
}

map<int, int> factorization(int n);
vector<int> factor(int n);
int remove_factors(int &n, vector<int> &factors, int prime);

int chinese_value(int a, int b, int r1, int r2) {
    int d = gcd(a, b);
    if (r1 % d != r2 % d) {
        return -1; // No solution
    }
    
    map<int, int> factorization_of_a = factorization(a);
    map<int, int> factorization_of_b = factorization(b);
    
    int a_prime = a, b_prime = b;
    
    for (auto &[prime, a_exponent] : factorization_of_a) {
        if (factorization_of_b.count(prime)) {
            int b_exponent = factorization_of_b[prime];
            if (a_exponent < b_exponent) {
                a_prime /= pow(prime, a_exponent);
            } else {
                b_prime /= pow(prime, b_exponent);
            }
        }
    }
    
    int first_coefficient = inverse(b_prime, a_prime);
    int second_coefficient = inverse(a_prime, b_prime);
    
    int solution = b_prime * first_coefficient * r1 + a_prime * second_coefficient * r2;
    int prime = a_prime * b_prime;
    
    solution = (solution % prime + prime) % prime;
    return solution;
}

map<int, int> factorization(int n) {
    vector<int> factors = factor(n);
    map<int, int> grouped_factors;
    for (int f : factors) {
        grouped_factors[f]++;
    }
    return grouped_factors;
}

vector<int> factor(int n) {
    vector<int> result;
    for (int prime : {2, 3, 5}) {
        remove_factors(n, result, prime);
    }
    int d = 0;
    int square_root = sqrt(n);
    while (n > 1 && d <= square_root) {
        for (int current_residue : {1, 7, 11, 13, 17, 19, 23, 29}) {
            int current_divisor = d + current_residue;
            if (current_divisor != 1 && n % current_divisor == 0) {
                remove_factors(n, result, current_divisor);
            }
        }
        d += 30;
    }
    if (n > 1) {
        result.push_back(n);
    }
    return result;
}

int remove_factors(int &n, vector<int> &factors, int prime) {
    while (n % prime == 0) {
        factors.push_back(prime);
        n /= prime;
    }
    return n;
}

std::pair<std::vector<int>, std::vector<std::vector<int>>> generate_table(int a, int b, int start_x, int start_y, int width, int height) {
    std::vector<std::vector<int>> modulo_table;
    std::vector<int> header;

    for (int j = start_x; j < start_x + width; ++j) {
        header.push_back(j);
    }

    for (int i = start_y; i < start_y + height; ++i) {
        std::vector<int> current_row;
        current_row.push_back(i);

        for (int j = start_x; j < start_x + width; ++j) {
            int result = chinese_value(a, b, i, j);
            current_row.push_back(result);
        }

        modulo_table.push_back(current_row);
    }

    return {header, modulo_table};
}

void print_table(const std::vector<int>& headers, const std::vector<std::vector<int>>& table) {
    // Print the header
    std::cout << std::setw(5) << " ";
    for (int h : headers) {
        std::cout << std::setw(5) << h;
    }
    std::cout << std::endl;

    // Print the table
    for (const auto& row : table) {
        for (const auto& cell : row) {
            if (cell != -1) {
                std::cout << std::setw(5) << cell;
            } else {
                std::cout << std::setw(5) << "";
            }
        }
        std::cout << std::endl;
    }
}

int main() {
    int a, b;
    std::cout << "Enter number a: ";
    std::cin >> a;
    std::cout << "Enter number b: ";
    std::cin >> b;

    if (std::cin.fail() || a <= 0 || b <= 0) {
        std::cout << "Please enter valid positive numbers for a and b." << std::endl;
        return 1;
    }

    int start_x = 0, start_y = 0;
    int width = b - start_x;
    int height = a - start_y;

    if (start_x < 0 || start_x >= b || start_y < 0 || start_y >= a ||
        width <= 0 || height <= 0 || (start_x + width - 1) >= b || (start_y + height - 1) >= a) {
        std::cout << "Invalid values for startX, startY, width, or height. Please check and try again." << std::endl;
        return 1;
    }

    std::cout << "Intersection of arithmetic progressions (mod " << a << ") and (mod " << b << ")" << std::endl;

    auto [headers, table] = generate_table(a, b, start_x, start_y, width, height);
    print_table(headers, table);

    return 0;
}
